src/lib/ledger_catchup/ledger_catchup.ml

```diff
@@ -25,47 +25,92 @@ module Make (Inputs : Inputs.S) :
   (* We would like the async scheduler to context switch between each iteration
      of external transitions when trying to build breadcrumb_path. Therefore, this 
      function needs to return a Deferred *)
-  let construct_breadcrumb_path ~logger initial_breadcrumb tree =
-    Rose_tree.Deferred.Or_error.fold_map tree
-      ~init:(Cached.pure initial_breadcrumb)
-      ~f:(fun cached_parent_breadcrumb cached_transition_with_hash ->
-        let open Deferred.Let_syntax in
-        let%map cached_result =
-          Cached.transform cached_transition_with_hash
-            ~f:(fun transition_with_hash ->
-              let open Deferred.Or_error.Let_syntax in
-              let parent_breadcrumb = Cached.peek cached_parent_breadcrumb in
-              let parent_state_hash =
-                Transition_frontier.Breadcrumb.transition_with_hash
-                  parent_breadcrumb
-                |> With_hash.hash
-              in
-              let current_state_hash = With_hash.hash transition_with_hash in
-              let%bind () =
-                Deferred.return
-                  (Result.ok_if_true
-                     ( State_hash.equal parent_state_hash
-                     @@ get_previous_state_hash transition_with_hash )
-                     ~error:
-                       ( Error.of_string
-                       @@ sprintf
-                            !"Previous external transition hash \
-                              %{sexp:State_hash.t} does not equal to current \
-                              external_transition's parent hash \
-                              %{sexp:State_hash.t}"
-                            parent_state_hash current_state_hash ))
-              in
-              let open Deferred.Let_syntax in
-              match%map
-                Transition_frontier.Breadcrumb.build ~logger
-                  ~parent:parent_breadcrumb ~transition_with_hash
-              with
-              | Ok new_breadcrumb -> Ok new_breadcrumb
-              | Error (`Fatal_error exn) -> Or_error.of_exn exn
-              | Error (`Validation_error error) -> Error error )
-          |> Cached.sequence_deferred
-        in
-        Cached.sequence_result cached_result )
+  let construct_breadcrumb_path ~logger frontier initial_state_hash tree =
+    (* If the breadcrumb we are targetting is removed from the transition
+     * frontier while we're catching up, it means this path is not on the
+     * critical path that has been chosen in the frontier. As such, we should
+     * drop it on the floor. *)
+    let breadcrumb_if_present () =
+      match Transition_frontier.find frontier initial_state_hash with
+      | None ->
+          let msg =
+            Printf.sprintf
```

Comment:
I think this should be a logged rather than Printf.

Reply:
This is logged, look at the next few lines